p2p/src/sync/peer.rs

```diff
@@ -13,56 +13,157 @@
+    id: ConstValue<PeerId>,
+    p2p_config: Arc<P2pConfig>,
+    chainstate_handle: subsystem::Handle<Box<dyn ChainstateInterface>>,
+    peer_manager_sender: UnboundedSender<PeerManagerEvent<T>>,
+    message_sender: UnboundedSender<(PeerId, SyncMessage)>,
+    events_receiver: UnboundedReceiver<PeerEvent>,
+    is_initial_block_download: Arc<AtomicBool>,
+    /// A list of headers received via the `HeaderListResponse` message that we haven't yet
+    /// requested the blocks for.
+    known_headers: Vec<BlockHeader>,
+    /// A list of blocks that we requested from this peer.
+    requested_blocks: BTreeSet<Id<Block>>,
+    /// A queue of the blocks requested this peer.
+    blocks_queue: VecDeque<Id<Block>>,
+}
+
+impl<T> Peer<T>
 where
     T: NetworkingService,
     T::SyncingMessagingHandle: SyncingMessagingService<T>,
 {
-    pub(super) async fn handle_message(
-        &mut self,
-        peer: PeerId,
-        message: SyncMessage,
-    ) -> Result<()> {
-        match message {
-            SyncMessage::HeaderListRequest(r) => {
-                self.handle_header_request(peer, r.into_locator()).await
+    pub fn new(
+        id: PeerId,
+        p2p_config: Arc<P2pConfig>,
+        chainstate_handle: subsystem::Handle<Box<dyn ChainstateInterface>>,
+        peer_manager_sender: UnboundedSender<PeerManagerEvent<T>>,
+        message_sender: UnboundedSender<(PeerId, SyncMessage)>,
+        events_receiver: UnboundedReceiver<PeerEvent>,
+        is_initial_block_download: Arc<AtomicBool>,
+    ) -> Self {
+        Self {
+            id: id.into(),
+            p2p_config,
+            chainstate_handle,
+            peer_manager_sender,
+            message_sender,
+            events_receiver,
+            is_initial_block_download,
+            known_headers: Vec::new(),
+            requested_blocks: BTreeSet::new(),
+            blocks_queue: VecDeque::new(),
+        }
+    }
+
+    /// Returns an identifier of the peer associated with this task.
+    pub fn id(&self) -> PeerId {
+        *self.id
+    }
+
+    pub async fn run(&mut self) -> Result<Void> {
+        self.request_headers().await?;
+
+        loop {
+            tokio::select! {
+                event = self.events_receiver.recv() => {
+                    let event = event.ok_or(P2pError::ChannelClosed)?;
+                    self.handle_event(event).await?;
+                },
+
+                block = async { self.blocks_queue.pop_front().expect("The block queue is empty") }, if !self.blocks_queue.is_empty() => {
+                    let res = self.send_block(block).await;
+                    self.handle_result(res).await?;
+                }
             }
-            SyncMessage::BlockListRequest(r) => {
-                self.handle_block_request(peer, r.into_block_ids()).await
+        }
+    }
+
+    async fn request_headers(&mut self) -> Result<()> {
+        let locator = self.chainstate_handle.call(|this| this.get_locator()).await??;
```

Comment:
Btw, one point to keep in mind that a locator is not an absolute value. I think it should also be possible to start from some point. Please check how bitcoin does it when negotiating headers. Something for another PR.

Reply:
https://github.com/mintlayer/mintlayer-core/issues/747